include/odp/api/spec/ipsec.h

```diff
@@ -712,10 +714,12 @@ typedef struct odp_ipsec_sa_param_t {
 	uint32_t context_len;
 
 	/** IPSEC SA direction dependent parameters */
-	union {
+	struct {
```

Comment:
Is this intentional?

Reply:
Yes it is, as explained in the github comment and the commit message.

Using a union has the drawback that default values cannot be defined for the union members as odp_ipsec_sa_param_init() cannot write them without affecting all the union members at the same time. And not being able to define defaults makes the _param_init() pattern used in ODP less useful. Changing the union to a struct fixes this.

Increased memory usage is a drawback but maybe it does not matter too much since this structure is needed only at SA creation time for parameter passing. In my x86 build the size seems to increase from 208 bytes to 272 bytes.

So, does the change make sense or should we just stick with the unions (and define fewer default values)?

Reply:
The only drawback is, sa_info is supposed to return this struct and the memory requirement would just increase. Or if we can state that odp_ipsec_sa_param_t.inbound/outbound would be ignored based on odp_ipsec_sa_param_t.dir, then it should be fine.

IOW, when dir = inbound, sa_info.param.outbound would have default values (& the other way round also). Would that be fine?

Reply:
I think the SA info API is a bit problematic as the current wording suggests that the returned param struct is a exact (?) copy of the one used in SA creation. But we do not want that as we do not want to force implementations to store the struct or the values of the ignored fields that cannot be recreated from internal SA data.

This problem exists already before the union-to-struct change, but the change makes it words.

I think I would like to propose the following change to the sa_info API, regardless of what we do with this PR:

```
 typedef struct odp_ipsec_sa_info_t {
-       /** Copy of IPSEC Security Association (SA) parameters */
+       /** IPsec SA parameters
+        *
+        * This is not necessarily an exact copy of the actual parameter
+        * structure used in SA creation. The fields that were relevant
+        * for the SA in the creation phase will have the same values,
+        * but other fields, such as tunnel parameters for a transport
+        * mode SA, will have undefined values.
+        */
        odp_ipsec_sa_param_t param;
```

Reply:
I agree to the suggestion on sa_info. With that change, move to struct is not an issue. Agreed that separate structs allows for specifying defaults better.